### PR TITLE
Legend Interface

### PR DESCRIPTION
**Pull request by lohani2280**
## Purpose
This patch implements a machine readable graph legend interface for GraphSpace web platform. The legend interface is highly interactive and comes up with a legend editor tool which allow users to add, modify, remove and save legend data for a graph using GUI. 

The users can also migrate some of the existing GraphSpace network having HTML legend format into corresponding JSON format at one-click of a button ([Demo](https://drive.google.com/open?id=1v6qAw9ruSWzBrMzH4eRpJuo4QsHI-v09))

### Feature Details
**Graph Legend Interface**

The Legend Interface is shown on top-right corner of the canvas for a graph in the `Graph visualization` Tab as shown below:
gs-screenshot-legend-interface

The interface has following features:
- It can be used to select nodes or edges of the graph corresponding to a legend just by clicking over that legend in the legend interface ([Demo](https://drive.google.com/open?id=1pUYS5CwgnL13NeoJG9xJlT9neuX7rd80))
- It can be dragged anywhere on top of the canvas for a graph ([Demo](https://drive.google.com/open?id=1H7NUNYnp3GxHUn2oCwAEeQd42ll5vHZ8))
- The Hide Legend/Show Legend button available in the side panel on the right side of the graph can be used to toggle between showing and hiding the Legend Interface ([Demo](https://drive.google.com/open?id=1OEcObQ1Ew_E0mZbcWoynkFnYdqExVlRl))

**Legend Editor Tool**

Legend Editor Tool provides features for users like:
- Edit a Legend label ([Demo](https://drive.google.com/open?id=1bJH21jal8UA52-sLlWuIMwP3pbyMhFZU))
- Remove a legend ([Demo](https://drive.google.com/open?id=1ry_t-uTi_esSQgRJRzOhQVlIjkaU_8kl))
- Add a node legend ([Demo](https://drive.google.com/open?id=1E-9o9XYU7w7CZRF95n1L4tgEQLO-yVNB))
- Add an edge legend ([Demo](https://drive.google.com/open?id=1h81eek_88EvFDy-r_dY5j1GxkuMaxi-K))
- Save legend ([Demo](https://drive.google.com/open?id=1U5dpwhBptlcREg88bimT4h5KIi_I3YuG) [Demo](https://drive.google.com/open?id=17FNAReKNunfE2a4isIfYdBP6urByDCTV))

## Approach
A Draggable DIV Element with id `cyLegendContainer` is created on the graph canvas and a legend graph is constructed over this DIV element from the legend data available in the style_json object using the cytoscape.js library . Now methods are implemented to make the DIV element collapsible and selectable(meaning whenever a user taps on the legend then corresponding elements in the graph gets selected). A Legend Editor Tool has been created to add, remove and update legend on this interface and finally save the updated legend data in the database. 

**New files added**
`templates/legend`: Contains HTML templates for Legend Editor Tool
`applications/legend_formatter.py`: Contains functions to convert a few GraphSpace network with HTML format legend to JSON format legend

### CODE REVIEW

File: static/css/graphspace.css [hunk 1]
```diff
@@ -128,16 +128,37 @@ div.container-fluid {
     position: absolute;
 }
 
-#cyGraphContainer {
+#graph{
```
Comment (adbharadwaj) [inline static/css/graphspace.css]: Lets not call it `graph`. It is very generic. ID should be descriptive. Call it `graphContainer`

File: graphspace/settings/local.py [hunk 2]
```diff
@@ -39,9 +39,9 @@
 DATABASES = {
     'default': {
         'ENGINE': 'django.db.backends.postgresql_psycopg2',
-        'NAME': 'test_database',
```
Comment (adbharadwaj) [inline graphspace/settings/local.py]: Remove this change from the pull request.

File: static/js/graphs_page.js [hunk 3]
```diff
@@ -991,6 +997,110 @@ var graphPage = {
         });
 
     },
+    constructLegend: function() {
```
Comment (adbharadwaj) [inline static/js/graphs_page.js]: Add some comments to describe the role of function.

File: templates/graph/index.html [hunk 4]
```diff
@@ -333,5 +343,21 @@ <h4 class="zero-margin d-inline-block">
...
         })
+
+        $("#legendBtn").click(function(e) {
+          var legendContainer = document.getElementById("cyLegendContainer");
+          var graphContainer = document.getElementById("cyGraphContainer");
+          if (legendContainer.style.display === "none") {
+            legendContainer.style.display = "block";
+            legendContainer.style.width = "20%";
+            graphContainer.style.width = "80%";
+            document.getElementById("legendBtn").innerHTML = "&#9776;" + " Hide Legends";
```
Comment (adbharadwaj) [inline templates/graph/index.html]: Remove the burger icon. Use `Legend` instead of `Legends`

File: graphspace/settings/local.py (hunk 2 above)
Comment (jahandaniyal) [reply graphspace/settings/local.py]: You can also add this file to .gitignore to avoid accidental inclusions.

File: applications/graphs/views.py [hunk 5]
```diff
@@ -494,6 +495,19 @@ def _update_graph(request, graph_id, graph={}):
 	authorization.validate(request, permission='GRAPH_UPDATE', graph_id=graph_id)
 	user_role = authorization.user_role(request)
 
+	if 'update_legend_format' in graph:
```
Comment (adbharadwaj) [inline applications/graphs/views.py]: Add comments!

File: applications/graphs/views.py [hunk 6]
```diff
@@ -494,6 +495,19 @@ def _update_graph(request, graph_id, graph={}):
...
+		if graph['update_legend_format'] == 1:
```
Comment (adbharadwaj) [inline applications/graphs/views.py]: use constant variables for 1.

File: applications/graphs/views.py [hunk 7]
```diff
@@ -494,6 +495,19 @@ def _update_graph(request, graph_id, graph={}):
...
+			graph = utils.serializer(graphs.get_graph_by_id(request, graph_id))
+			style_json = convert_html_legend_1(graph['graph_json'], graph['style_json'])
+			del graph['graph_json']["data"]["description"]
+			return utils.serializer(graphs.update_graph(request, graph_id=graph_id, style_json=style_json, graph_json=graph['graph_json']))
+
+		if graph['update_legend_format'] == 2:
```
Comment (adbharadwaj) [inline applications/graphs/views.py]: Use named constant variables.

File: applications/graphs/views.py (hunk 6 above)
Comment (adbharadwaj) [reply applications/graphs/views.py]: `HTML_TABLE_FORMAT_1`

File: applications/graphs/views.py (hunk 7 above)
Comment (adbharadwaj) [reply applications/graphs/views.py]: `HTML_TABLE_FORMAT_2`

File: applications/graphs/views.py (hunk 5 above)
Comment (adbharadwaj) [inline applications/graphs/views.py]: Move this logic to controller.

Comment (adbharadwaj) [reply applications/graphs/views.py]: Move this if logic to one function `convert_html_legend`

File: graphspace/formatter/legend_formatter.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
+import json
+import lxml.html
+
+
+def filter_legend_data(legend_desc_shape, noise):
```
Comment (adbharadwaj) [inline graphspace/formatter/legend_formatter.py]: Add function documentation.

File: graphspace/formatter/legend_formatter.py [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+	filtered_legend = [x for x in legend_desc_shape if x not in noise]
+	return filtered_legend
+
+def find_color(root, node_legend_length='none', edge_legend_length='none'):
```
Comment (adbharadwaj) [inline graphspace/formatter/legend_formatter.py]: Add function documentation.

Comment (adbharadwaj) [reply graphspace/formatter/legend_formatter.py]: Why `none` as default?

File: graphspace/formatter/legend_formatter.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+	count = 0
+	for element in root.iter('font'):
+		count += 1
+		x = element.attrib['color'].replace('\"','')
+		color_shape.append(x.replace('\\',''))
+		if(count == node_legend_length + edge_legend_length):
+			return color_shape
+
+def convert_unicode_node_shape(node_shape):
+	acceptedShapes = {u'\u25a0':"rectangle", u'\u25b2':"triangle", u'\u25cf':"circle", u'\u2605': "star", u'\u25fe': "square"}
```
Comment (adbharadwaj) [inline graphspace/formatter/legend_formatter.py]: define constants in caps.

Comment (adbharadwaj) [reply graphspace/formatter/legend_formatter.py]: change to snake case.

File: graphspace/formatter/legend_formatter.py [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+
+	elements_color = find_color(root, len(node_shape), len(edge_shape))
+	node_color = elements_color[:len(node_shape)]
+	edge_color = elements_color[len(edge_shape)-1:]
+	node_shape = convert_unicode_node_shape(node_shape)
+	edge_shape = convert_unicode_edge_shape(edge_shape)
+
+
+	legend_json = {}
+	legend_json["legend"] = {}
```
Comment (adbharadwaj) [inline graphspace/formatter/legend_formatter.py]: use single qoutes.

File: graphspace/formatter/legend_formatter.py [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+
+def convert_unicode_edge_shape(edge_shape):
+	acceptedShapes  = {u'\u27f6':"triangle", u'\u22a3':"tee", u'\u2014':"none"}
+	new_edge_shape = []
+	for x in edge_shape:
+		new_edge_shape.append(acceptedShapes[x])
+	return new_edge_shape
+
+def convert_html_legend_1(graph_json,  style_json):
+	html_table_string = json.dumps(graph_json["data"]['description'])
```
Comment (adbharadwaj) [inline graphspace/formatter/legend_formatter.py]: Add an url to an example public graph to the documentation for reference.

File: static/js/graphs_page.js [hunk 13]
```diff
@@ -437,10 +438,82 @@ var graphPage = {
...
 
+        if('legend' in style_json){
+            graphPage.cyLegend = graphPage.constructLegend(style_json);
+        }
+
         graphPage.cyGraph.panzoom();
 
         utils.initializeTabs();
 
+        var update_legend_format = 0;
```
Comment (adbharadwaj) [inline static/js/graphs_page.js]: define a constant for 0.

File: static/js/graphs_page.js [hunk 14]
```diff
@@ -437,10 +438,82 @@ var graphPage = {
...
+        $("#graphDetailsTabBtn").click(function () {
+            if(graph_json['data']['description']) {
+
+                var html_legend_data = graph_json['data']['description'];
+                html_legend_data = $($.parseHTML(html_legend_data));
+
+                try {
```
Comment (adbharadwaj) [inline static/js/graphs_page.js]: Refactor the logic to small function. Move the logic to a module specific to legend.

File: static/js/graphs_page.js [hunk 15]
```diff
@@ -455,15 +528,39 @@ var graphPage = {
             graphPage.saveLayout($('#saveLayoutNameInput').val(), '#saveLayoutModal');
         });
 
+       $('#saveOnExitLegendBtn').click(function () {
```
Comment (adbharadwaj) [inline static/js/graphs_page.js]: Move to `graphsPage.legend.init()`

File: static/js/graphs_page.js [hunk 16]
```diff
@@ -455,15 +528,39 @@ var graphPage = {
...
+            if('legend' in style_json){
+                graphPage.legendEditor.removeBinLegend();
+            }
+        });
+
         $('#layoutEditorBtn').click(function () {
             graphPage.layoutEditor.init();
         });
 
+        $('#legendEditorBtn').click(function () {
```
Comment (adbharadwaj) [inline static/js/graphs_page.js]: @adbharadwaj
adbharadwaj now Member

Move to `graphsPage.legend.init()`

File: static/js/graphs_page.js [hunk 17]
```diff
@@ -455,15 +528,39 @@ var graphPage = {
...
+                graphPage.legendEditor.init();
+            }
+        });
+
         $('#exitLayoutEditorBtn').click(function () {
             $('#exitLayoutBtn').removeClass('hidden');
             $('#saveOnExitLayoutModal').modal('show');
         });
 
+       $('#exitLegendEditorBtn').click(function () {
```
Comment (adbharadwaj) [inline static/js/graphs_page.js]: @adbharadwaj
adbharadwaj now Member

Move to `graphsPage.legend.init()`

File: static/js/graphs_page.js [hunk 18]
```diff
@@ -455,15 +528,39 @@ var graphPage = {
...
+            $('#exitLegendBtn').removeClass('hidden');
+            $('#saveOnExitLegendModal').modal('show');
+        });
+
+        $('#exitLegendBtn').click(function () {
```
Comment (adbharadwaj) [inline static/js/graphs_page.js]: Move to `graphsPage.legend.init()`

File: static/js/graphs_page.js [hunk 19]
```diff
@@ -483,10 +580,16 @@ var graphPage = {
         this.filterNodesEdges.init();
         this.colaLayoutWidget.init();
 
-        if (window.location.hash == '#editor') {
+        if (window.location.hash == '#layouteditor') {
```
Comment (adbharadwaj) [inline static/js/graphs_page.js]: Use `layoutEditor`

File: static/js/graphs_page.js [hunk 20]
```diff
@@ -483,10 +580,16 @@ var graphPage = {
...
             $('#layoutEditorBtn').trigger('click');
         }
 
+        if (window.location.hash == '#legendeditor') {
```
Comment (adbharadwaj) [inline static/js/graphs_page.js]: Use `legendEditor`

File: static/js/graphs_page.js [hunk 21]
```diff
@@ -638,8 +746,7 @@ var graphPage = {
             }
 
             positions_json = cytoscapeGraph.getNodePositions(graphPage.cyGraph);
-            style_json = cytoscapeGraph.getStylesheet(graphPage.cyGraph);
```
Comment (adbharadwaj) [inline static/js/graphs_page.js]: undo the change

File: static/js/graphs_page.js [hunk 22]
```diff
@@ -991,6 +1098,119 @@ var graphPage = {
         });
 
     },
+    constructLegend: function(styleJSON) {
```
Comment (adbharadwaj) [inline static/js/graphs_page.js]: Move to `graphsPage.legend`

File: static/js/graphs_page.js [hunk 23]
```diff
@@ -991,6 +1098,119 @@ var graphPage = {
...
+               'line-style': edge_legend[Object.keys(edge_legend)[i]]['line-style'],
+               'target-arrow-shape': edge_legend[Object.keys(edge_legend)[i]]['arrow-shape'],
+               'target-arrow-color': edge_legend[Object.keys(edge_legend)[i]]['line-color']
+            });
+
+            k=k+40;
+        }
+        return cy;
+    },
+    onTapLegendGraphElement: function(evt) {
```
Comment (adbharadwaj) [inline static/js/graphs_page.js]: Move to `graphsPage.legend`

File: static/js/graphs_page.js [hunk 24]
```diff
@@ -1357,6 +1577,240 @@ var graphPage = {
             );
         }
     },
+    legendEditor: {
```
Comment (adbharadwaj) [inline static/js/graphs_page.js]: Move to `graphsPage.legend`

File: static/js/graphs_page.js [hunk 25]
```diff
@@ -1357,6 +1577,240 @@ var graphPage = {
...
+                    'background-image': '../images/trash.png',
+                    'background-color': 'white',
+                    'width':'20px',
+                    'height':'25px',
+                    'shape': 'rectangle'
+                });
+                k=k+40;
+            }
+        },
+        rearrangeLegend: function(bin_pos_y) {
```
Comment (adbharadwaj) [inline static/js/graphs_page.js]: Document each function.

File: static/js/graphs_page.js [hunk 26]
```diff
@@ -1357,6 +1577,240 @@ var graphPage = {
...
+                style_json['legend']['edges'][e_label]['line-color'] = e_color;
+                style_json['legend']['edges'][e_label]['arrow-shape'] = e_arrow_shape;
+
+                graphPage.cyLegend.remove(graphPage.cyLegend.elements());
+                graphPage.cyLegend=graphPage.constructLegend(style_json);
+                graphPage.legendEditor.init();
+            },
+            close: function() {
+                if(save) {
+                    console.log(graphPage.legendEditor.edgeLegendEditor.edge_legend);
```
Comment (adbharadwaj) [inline static/js/graphs_page.js]: comment all logs

File: templates/graph/default_sidebar.html [hunk 27]
```diff
@@ -42,9 +51,15 @@
             </a>
         </li>
         <li>
-            <a id="layoutEditorBtn" class="btn sidebar-nav-pills" href="#editor" data-target="#layoutEditorSideBar">
+            <a id="layoutEditorBtn" class="btn sidebar-nav-pills" href="#layouteditor" data-target="#layoutEditorSideBar">
```
Comment (adbharadwaj) [inline templates/graph/default_sidebar.html]: layoutEditor

File: templates/graph/default_sidebar.html [hunk 28]
```diff
@@ -42,9 +51,15 @@
...
                 <i class="fa fa-pencil-square-o fa-lg"></i> Use Layout <br> Editor
             </a>
+
+        </li>
+        <li>
+            <a id="legendEditorBtn" class="btn sidebar-nav-pills" href="#legendeditor" data-target="#legendEditorSideBar">
```
Comment (adbharadwaj) [inline templates/graph/default_sidebar.html]: legendEditor

File: templates/graph/legend_editor_sidebar.html [hunk 29]
```diff
@@ -0,0 +1,17 @@
+-<!-- sidebar -->
```
Comment (adbharadwaj) [inline templates/graph/legend_editor_sidebar.html]: Move to `templates/graph/legend/`

File: applications/graphs/views.py (hunk 5 above)
Comment (jahandaniyal) [inline applications/graphs/views.py]: Views should only do basic computations - passing request to the correct methods in controllers.py and returning a HTTPResponse. 
I believe a custom method like `graphs.update_graph_with_legend()` which does all the computation and returns a Graph dataset will make more sense.

Then this view could be reduced to -
```
if 'update_legend_format' in graph:
        return utils.serializer(graphs.update_graph_with_legend(request, graph_id=graph_id))
```

File: graphspace/formatter/legend_formatter.py [hunk 30]
```diff
@@ -0,0 +1,107 @@
...
+	legend_json["legend"]["nodes"] = {};
+	legend_json["legend"]["edges"] = {};
+
+	for i in range(len(node_desc)):
+		legend_json["legend"]["nodes"][node_desc[i]] = {};
+		legend_json["legend"]["nodes"][node_desc[i]]["background-color"] =  node_color[i];
+		legend_json["legend"]["nodes"][node_desc[i]]["shape"] =  'circle';
+
+	style_json.update(legend_json)
+	return style_json
```
Comment (jahandaniyal) [inline graphspace/formatter/legend_formatter.py]: Add newline.
https://www.python.org/dev/peps/pep-0008/#id21

**Comment by jahandaniyal** [comment]
Hi @lohani2280 - You may also want to avoid **force pushes**, it's alright for a branch where only you are working but it still has some pitfalls.
I suggest this blog, it's a good read - https://blog.developer.atlassian.com/force-with-lease/

**Comment by lohani2280** [comment]
> Hi @lohani2280 - You may also want to avoid **force pushes**, it's alright for a branch where only you are working but it still has some pitfalls.
> I suggest this blog, it's a good read - https://blog.developer.atlassian.com/force-with-lease/

Hey. Thanks @jahandaniyal. Will look into it.

**Comment by jahandaniyal** [comment]
@adbharadwaj - I was also wondering if @lohani2280 has been briefed about [Graphspace Manual](https://github.com/adbharadwaj/graphspace-manual).  I have my share of experience working with the manual and I can help with any queries he might have.

File: applications/legend_formatter.py [hunk 31]
```diff
@@ -0,0 +1,111 @@
...
+		color_shape.append(x.replace('\\',''))
+		if(count == node_legend_length + edge_legend_length):
+			return color_shape
+
+def convert_unicode_node_shape(node_shape):
+	ACCEPTED_NODE_SHAPES = {u'\u25a0':"rectangle", u'\u25b2':"triangle", u'\u25cf':"circle", u'\u2605': "star", u'\u25fe': "square"}
+	new_node_shape = []
+	for x in node_shape:
+		new_node_shape.append(ACCEPTED_NODE_SHAPES[x])
+	return new_node_shape
```
Comment (jahandaniyal) [inline applications/legend_formatter.py]: I would suggest using List Comprehension for this case  :- 
`return [ACCEPTED_NODE_SHAPES[x] for x in node_shape]`

File: applications/legend_formatter.py [hunk 32]
```diff
@@ -0,0 +1,111 @@
...
+
+def convert_unicode_edge_shape(edge_shape):
+	ACCEPTED_EDGE_SHAPES  = {u'\u27f6':"triangle", u'\u22a3':"tee", u'\u2014':"none"}
+	new_edge_shape = []
+	for x in edge_shape:
+		new_edge_shape.append(ACCEPTED_EDGE_SHAPES[x])
+	return new_edge_shape
```
Comment (jahandaniyal) [inline applications/legend_formatter.py]: Same as above.

File: applications/legend_formatter.py [hunk 33]
```diff
@@ -0,0 +1,111 @@
...
+	node_color = elements_color[:len(node_shape)]
+	edge_color = elements_color[len(edge_shape)-1:]
+	node_shape = convert_unicode_node_shape(node_shape)
+	edge_shape = convert_unicode_edge_shape(edge_shape)
+
+	print(elements_color)
+	print(node_color)
+	print(edge_color)
+	print(node_shape)
+	print(edge_shape)
```
Comment (jahandaniyal) [inline applications/legend_formatter.py]: Unless it's put here by design, you should remove these from the final version!

File: applications/legend_formatter.py [hunk 34]
```diff
@@ -0,0 +1,111 @@
...
+def convert_html_legend_2(graph_json,  style_json):
+	html_table_string = graph_json['data']['description']
+	root = lxml.html.fromstring(html_table_string)
+
+	node_shape_desc = root.xpath('//td//text()');
+	node_shape = [node_shape_desc[0],node_shape_desc[2]]
+
+	node_color = []
+	for element in root.iter('td'):
+		try:
```
Comment (jahandaniyal) [inline applications/legend_formatter.py]: Which cases lead to exception here?

File: applications/legend_formatter.py [hunk 35]
```diff
@@ -0,0 +1,111 @@
...
+	print(elements_color)
+	print(node_color)
+	print(edge_color)
+	print(node_shape)
+	print(edge_shape)
+
+	legend_json = {}
+	legend_json['legend'] = {}
+	legend_json['legend']['nodes'] = {}
+	legend_json['legend']['edges'] = {}
```
Comment (jahandaniyal) [inline applications/legend_formatter.py]: Probably an anti-pattern (nitpicking) - Though this seems okay to me. 
This `legend_json = {'legend': { 'nodes':{}, 'edges':{} } }`  or perhaps use defaultdict. I would let @adbharadwaj decide what he prefers in Graphspace.
I would rate it as unimportant in any case.

File: static/js/graphs_page.js [hunk 36]
```diff
@@ -2654,6 +2665,57 @@ var graphPage = {
...
+                var link = $('<a>').attr('href', file).attr('download', filename + '.' + fileformat);
+                $('body').append(link);
+                link.get(0).click();
+                link.get(0).remove();
+            }
+
+            if (fileformat === 'jpg') {
+                var cyGraph_b64 = graphPage.cyGraph.jpeg();
+                var cyLegend_b64 = graphPage.legend.cyLegend.jpeg();
+                cyGraphImg.onload = function() {
```
Comment (adbharadwaj) [inline static/js/graphs_page.js]:
```js
var cyGraphImgPromise, cyLegendImgPromise;

                cyGraphImg.onload = function() {
                    canvas.width = cyGraphImg.width;
                    canvas.height = cyGraphImg.height;
                    cyLegendImg.src = cyLegend_b64;
                    cyGraphImgPromise = Promise.resolve('Loaded cyGraphImg!');
                };
                cyLegendImg.onload = function() {
                    context.globalAlpha = 1;
                    context.drawImage(cyGraphImg, 0, 0);
                    context.globalAlpha = 1;
                    context.drawImage(cyLegendImg, 0, 0);
                    cyLegendImgPromise = Promise.resolve('Loaded cyLegendImg!');
                };

                
                Promise.all([cyGraphImgPromise, cyLegendImgPromise]).then(function(values) {
                    cyGraphImg.src = cyGraph_b64;
                    var file  = canvas.toDataURL("image/jpeg");
                    var link = $('<a>').attr('href', file).attr('download', filename + '.' + fileformat);
                    $('body').append(link);
                    link.get(0).click();
                    link.get(0).remove();
                });
```
